flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadSchedulerTest.java

```diff
@@ -192,6 +194,53 @@ public void testOnReadBufferRequestError() throws Exception {
         assertAllResourcesReleased();
     }
 
+    @Test
+    public void testIncreaseDeadlineWhenRecycleBuffer() throws Exception {
+        testRequestDeadlineInternal(true);
+        assertTrue(readScheduler.getBufferRequestDeadline().hasTimeLeft());
+    }
+
+    @Test
+    public void testDeadlineTimeoutWhenNoRecycleBuffer() throws Exception {
+        testRequestDeadlineInternal(false);
+        assertFalse(readScheduler.getBufferRequestDeadline().hasTimeLeft());
+    }
+
+    private void testRequestDeadlineInternal(boolean needTriggerRecycleBuffer) throws Exception {
+        readScheduler.setRequestBufferDeadline(Deadline.fromNow(Duration.ofMillis(500)));
+        assertTrue(readScheduler.getBufferRequestDeadline().hasTimeLeft());
+
+        readScheduler.run();
+        // Waiting for the deadline timeout
+        Thread.sleep(800);
+
+        if (needTriggerRecycleBuffer) {
+            // If at least one buffer is recycled, the deadline will be increased
+            triggerRecycleBuffer();
+        }
+    }
+
+    private void triggerRecycleBuffer() throws Exception {
```

Comment:
We can just request and recycle buffers directly to the buffer pool which can make the logic less complicated.

Reply:
Thanks again for these comments.
 I have updated the code. Could you please help check it again? @wsry